### PR TITLE
[ADVAPI32] Unify vista registry sources

### PR DESCRIPTION
**Pull request by tkreuzer**
## Purpose

This unifies the sources for vista registry functions and removes duplicated code.

## Proposed changes
- Fix bugs in RegpCopyTree
- In advapi32_vista compile the code from advapi32/reg/reg.c
- Export RegCopyTreeW from advapi32_vista
- Link userenv and services to advapi32_vista
- Remove duplicated code / hacks from advap32_vista, userenv and services

## Tests
- ✅ GCCLin_x86 on Test KVM: https://reactos.org/testman/compare.php?ids=92099,92105,92110,92121
- ✅ MSVC_x64 on Test KVM_x64: https://reactos.org/testman/compare.php?ids=90726,90727,90733,90739

### CODE REVIEW

File: dll/win32/advapi32/reg/reg.c [hunk 1]
```diff
@@ -5129,4 +5138,194 @@ RegUnLoadKeyW(HKEY hKey,
...
+    return result;
+}
+
+/************************************************************************
+ *  RegLoadMUIStringA
+ *
+ * @implemented
+ */
+LONG WINAPI
+RegLoadMUIStringA(IN HKEY hKey,
```
Comment (HBelusca) [inline dll/win32/advapi32/reg/reg.c:5291]:
```suggestion
RegLoadMUIStringA(
    IN HKEY hKey,
```

File: dll/win32/advapi32/reg/reg.c [hunk 2]
```diff
@@ -5129,4 +5138,194 @@ RegUnLoadKeyW(HKEY hKey,
...
+    return cMaxChars;
+}
+
+/************************************************************************
+ *  RegLoadMUIStringW
+ *
+ * @implemented
+ */
+LONG WINAPI
+RegLoadMUIStringW(IN HKEY hKey,
```
Comment (HBelusca) [inline dll/win32/advapi32/reg/reg.c:5192]:
```suggestion
RegLoadMUIStringW(
    IN HKEY hKey,
```

**Review by HBelusca — APPROVED** [review]

**Comment by HBelusca** [comment]
In commit log of second commit, the following line:
> - Remove duplicated code / hacks from advap32_vista, userenv and services

should read:
> - Remove duplicated code / hacks from advapi32_vista, userenv and services

(typo in the module name).

File: dll/win32/advapi32/reg/reg.c [hunk 3]
```diff
@@ -5129,4 +5138,194 @@ RegUnLoadKeyW(HKEY hKey,
...
+    while (idxString--) pString += *pString + 1;
+
+    /* If no buffer is given, return length of the string. */
+    if (!pwszBuffer) return *pString;
+
+    /* Else copy over the string, respecting the buffer size. */
+    cMaxChars = (*pString < cMaxChars) ? *pString : (cMaxChars - 1);
+    if (cMaxChars >= 0)
+    {
+        memcpy(pwszBuffer, pString + 1, cMaxChars * sizeof(WCHAR));
```
Comment (GeoB99) [inline dll/win32/advapi32/reg/reg.c:5179]: Maybe use `RtlCopyMemory` (and in the other place in `RegLoadMUIStringW` too)?

File: dll/win32/advapi32/reg/reg.c [hunk 4]
```diff
@@ -5129,4 +5138,194 @@ RegUnLoadKeyW(HKEY hKey,
...
+        pwszExpandedBuffer = HeapAlloc(GetProcessHeap(), 0, cbData);
+        memcpy(pwszExpandedBuffer, pwszTempBuffer, cbData);
+    }
+
+    /* If the value references a resource based string, parse the value and load the string.
+     * Else just copy over the original value. */
+    result = ERROR_SUCCESS;
+    if (*pwszExpandedBuffer != L'@') /* '@' is the prefix for resource based string entries. */
+    {
+        lstrcpynW(pszOutBuf, pwszExpandedBuffer, cbOutBuf / sizeof(WCHAR));
```
Comment (GeoB99) [inline dll/win32/advapi32/reg/reg.c:5255]: Use `StringCchCopyW` instead?

File: dll/win32/advapi32/reg/reg.c (hunk 3 above)
Comment (tkreuzer) [reply dll/win32/advapi32/reg/reg.c:5179]: This code is simply moved from a different place and I'm not gonna change it for some style improvements.

File: dll/win32/advapi32/reg/reg.c [hunk 5]
```diff
@@ -5129,4 +5138,194 @@ RegUnLoadKeyW(HKEY hKey,
...
+
+        /* Format of the expanded value is 'path_to_dll,-resId' */
+        if (!pComma || pComma[1] != L'-')
+        {
+            result = ERROR_BADKEY;
+            goto cleanup;
+        }
+
+        uiStringId = _wtoi(pComma + 2);
+        *pComma = L'\0';
```
Comment (katahiromz) [inline dll/win32/advapi32/reg/reg.c:5272]:
```suggestion
        *pComma = UNICODE_NULL;
```

**Review by learn-more — APPROVED** [review]

**Comment by tkreuzer** [comment]
Looks like the failures are random. We really need to fix those ransom test failures to make testing less of a PITA.

**Pull request merged**